crates/nu-cli/src/commands/math/product.rs

```diff
@@ -0,0 +1,128 @@
+use crate::commands::WholeStreamCommand;
+use crate::prelude::*;
+use nu_errors::ShellError;
+use nu_protocol::{
+    hir::{convert_number_to_u64, Number},
+    Primitive, Signature, UntaggedValue, Value,
+};
+
+pub struct SubCommand;
+
+#[async_trait]
+impl WholeStreamCommand for SubCommand {
+    fn name(&self) -> &str {
+        "math product"
+    }
+
+    fn signature(&self) -> Signature {
+        Signature::build("math product")
+    }
+
+    fn usage(&self) -> &str {
+        "Finds the product of a list of numbers or tables"
+    }
+
+    async fn run(
+        &self,
+        args: CommandArgs,
+        registry: &CommandRegistry,
+    ) -> Result<OutputStream, ShellError> {
+        run_with_function(
+            RunnableContext {
+                input: args.input,
+                registry: registry.clone(),
+                shell_manager: args.shell_manager,
+                host: args.host,
+                ctrl_c: args.ctrl_c,
+                current_errors: args.current_errors,
+                name: args.call_info.name_tag,
+                raw_input: args.raw_input,
+            },
+            product,
+        )
+        .await
+    }
+
+    fn examples(&self) -> Vec<Example> {
+        vec![Example {
+            description: "Get the product of a list of numbers",
+            example: "echo [2 3 3 4] | math product",
+            result: Some(vec![UntaggedValue::int(72).into()]),
+        }]
+    }
+}
+
+fn to_byte(value: &Value) -> Option<Value> {
+    match &value.value {
+        UntaggedValue::Primitive(Primitive::Int(num)) => Some(
+            UntaggedValue::Primitive(Primitive::Filesize(convert_number_to_u64(&Number::Int(
+                num.clone(),
+            ))))
+            .into_untagged_value(),
+        ),
+        _ => None,
+    }
+}
+
+/// Calculate product of given values
+pub fn product(values: &[Value], name: &Tag) -> Result<Value, ShellError> {
+    let prod = reducer_for(Reduce::Product);
+
+    let first = values.get(0).ok_or_else(|| {
+        ShellError::unexpected("Cannot perform aggregate math operation on empty data")
+    })?;
+
+    match first {
+        v if v.is_filesize() => to_byte(&prod(
+            UntaggedValue::int(1).into_untagged_value(),
+            values
+                .to_vec()
+                .iter()
+                .map(|v| match v {
+                    Value {
+                        value: UntaggedValue::Primitive(Primitive::Filesize(num)),
+                        ..
+                    } => UntaggedValue::int(*num as usize).into_untagged_value(),
+                    other => other.clone(),
+                })
+                .collect::<Vec<_>>(),
+        )?)
+        .ok_or_else(|| {
+            ShellError::labeled_error(
+                "could not convert to big decimal",
+                "could not convert to big decimal",
+                &name.span,
+            )
+        }),
+
+        v if v.is_none() => prod(
```

Comment:
Yes, so I pretty much followed `math sum` implementation and didn't think of the implications of this (given my naivety with the codebase). Since I first did pass on `math product`, looks like a lot of the code has been refactored and I was just trying to achieve parity with `math sum`. 

As you mentioned, would it make more sense to throw error instead of interpreting `Primitive::Nothing`?

Reply:
> As you mentioned, would it make more sense to throw error instead of interpreting `Primitive::Nothing`?

Yeah, I think we want to lean towards this approach. Again, don't worry about it in this PR 🙂 

The reason for it is that we could silently ignore a `Nothing` in a stream that the end user doesn't expect. In that scenario, our math functions would happily still produce a result, even though it's wrong.

I'd love to make it so that we can be explicit in that kind of way, but still allow our users to override that behaviour. For example (this won't work today):

```
alias math sum [] { default column 0 | math sum }
```